### PR TITLE
Brian dragons

### PR DESCRIPTION
**Pull request by bmerino95**
Adding DRAGONS notebooks for the reduction of imaging data with GMOS, GNIRS, NIRI, GSAOI, and FLAMINGOS-2 data.

### CODE REVIEW

**Review by jacquesalice — CHANGES_REQUESTED** [review]
Hi @bmerino95 thank you for these! Here are my comments. The first section refers to the directory structure and things that are applicable to all 6 notebooks. The sections after that are comments specific to some of the notebooks.

**General**
1. Rename the directory "DRAGONS_notebooks" --> "DRAGONSExamples" (or something like this)
2. Since users will be using the DRAGONS (Py3.7) kernel for these notebooks, they shouldn't have to `conda install` anything. So you can remove the "Before you try the notebooks" section in the README as well as the "don't forget to run conda install..." message at the top of each notebook 
3. Add your name to the list of authors!!
4. In the "Downloading the data" sections, you don't need to use wget for each file. Instead, do this:
```
%%bash

# # Science 
echo "\
http://archive.gemini.edu/file/S20131121S0075.fits
http://archive.gemini.edu/file/S20131121S0076.fits
http://archive.gemini.edu/file/S20131121S0077.fits
http://archive.gemini.edu/file/S20131121S0078.fits
http://archive.gemini.edu/file/S20131121S0079.fits
http://archive.gemini.edu/file/S20131121S0080.fits
http://archive.gemini.edu/file/S20131121S0081.fits
http://archive.gemini.edu/file/S20131121S0082.fits
http://archive.gemini.edu/file/S20131121S0083.fits\
" > science.list

# download with wget FITS files from Gemini archive for processing (in quiet mode)
wget --no-check-certificate -N -q -i science.list
```
5. We've found that doing `mkdir` in our notebook collection doesn't always work the way it's supposed to. Remove the cells that make a directory called raw and just keep the FITS files in the same directory as the notebook
6. In the "Clean up" sections, change what's currently there to:
```
#%%bash
#
#rm -f *.fits *.list *.log # remove files
#rm -rf calibrations/ # remove directories
```
7. In the imports section, remove "** (you'll probably have to install the wget and ipympl libraries)" from the title
8. In the imports section, remove `import wget`
9. In the "Create file lists" section, when you say "refer to the dataselect documentation for details..." can you include a link to that documentation?
10. If possible, make the image plot in RA/Dec rather than pixels
11. Add axes titles to image plot
12. In the "Create file lists" section, when you say "See the Astrodata User Manual for information about Astrodata" can you include a link to that user manual?

**F2_IM_WISEJ0413-4750.ipynb**
1. In the "About the dataset" section, replace "observced" --> "observed"
2. In the "Downloading the data" section, you don't need to download the .tar file, so remove that cell

**GSAOI_NGC5128.ipynb**
1. In the "About the dataset" section, can you add a bit more info about GSAOI? Similar to the info you have about the other datasets in the other notebooks
2. In the imports section, remove the part about having to install disco_stu (since it is not used anywhere in the notebook)
3. In the "Downloading the data" section, you don't need to download the .tar file, so remove that cell

**NIRI_SN2014J.ipynb**
1. In the "Downloading the data" section, you don't need to download the .tar file, so remove that cell

**Comment by galaxyumi** [comment]
@bmerino95 Are you done with implementing Alice's comments? Please let me know if your NBs are ready to review. :)

**Comment by bmerino95** [comment]
> @bmerino95 Are you done with implementing Alice's comments? Please let me know if your NBs are ready to review. :)

Hi Yumi! Thank you for checking in. I still have a few small things that I need to add, but it is almost ready.

**Comment by bmerino95** [comment]
Hello @jacquesalice and @galaxyumi! All of the DRAGONS notebooks have been revised and are ready for review. Please let me know if you need help running them or want me to make any changes. :)

**Comment by jacquesalice** [comment]
Thanks @bmerino95 ! Yumi is at a conference this week so won't be able to look at them until next week

**Comment by galaxyumi** [comment]
@bmerino95 Hi Brian, sorry for late feedback. I finally managed to review all 7 NBs. Nice job! 

Here are general comments that apply to all (or most of) the NBs:
1. Please set the NB default kernel to DRAGONS (Py3.7) and save it, if possible. 
2. Please tailor README.md to the DL version, and make sure to include information about the custom kernel DRAGONS (Py3.7). 
3. The full URLs in Summary could be removed by using the href attribute. Insert the full url to the ‘here’ text, allowing the users just to click ‘here’ to go to the destinations.
4. Please delete commented items in the “Importing Python Libraries” sections. By the way, do you need "%matplotlib inline"?
5. It would be helpful if you can add a few describing sentences for the "Downloading the data" sections. For example, "Downloading Y-band images from the Gemini archive to the current working directory. This step needs to be excused once." 
6. Please keep the consistency in the style of section titles - some have only the first word capitalized, while others capitalize every word.
7. Please don’t forget to export the rendered version to the HTML format once you finalize your NBs.
8. Sub-directory names and NB filenames could be more descriptive/informative replacing the specific object names with source types. If you make these changes, please don’t forget to update the README file.
Example suggestions are:
F2_IM_WISEJ0413-4750.ipynb -> Flamingos2_Imaging_BrownDwarf.ipynb
GMOS_IM_FIELD.ipynb -> GMOS_Imaging_StellarField.ipynb
9. Downloading the data section
Please doubly comment out (i.e., ## , not a single #) actual comments to make sure they are not uncommented when the user uncomment lines that actually need to be uncommented and executed. Or another way you can consider is to set a boolean flag, e.g., download_data = True. **IF** the flag is set, then download the data. **Else**, just print out something like “Skip data download” or simply skip the part. This way, the users only need to deal with that flag, and don’t need to worry about commenting out/uncommenting the cell manually every time . This could be also applicable to the Optional section where removing files.
10. In multiple NBs, I noticed the following message under the Create Master Dark or Bias sections: “Ginga not installed, use other viewer, or no viewer”. Is this expected?
11. In multiple NBs, I noticed unused packages, modules, functions are imported. For example, `‘import numpy as np’, ‘import matplotlib.cbook as cbook’.` Please double check to see if you import unused stuff.
12. Quotation mark looks different than the typical one - look at the words “VAR”, “DQ” as examples. It’s not opening and closing, but is slanted in the same direction. 

Here are some comments that are specific to individual NBs:
1. GMOS_Is_J2145+0031.ipynb
- In the Processed Arc - Wavelength Solution/ Processed Standard - Sensitivity Function/ Science Observations sections, I got a long, but repeated warning message as follows: 
`“WARNING: AstropyDeprecationWarning: ASDF functionality for astropy is being moved out of the astropy package to the new asdf-astropy package. Please use this package instead. astropy.io.misc.asdf is deprecated since astropy 5.1 and will be removed in a future release. [astropy.io.misc.asdf.types]”. `
If you want to let it print out, please put a sentence about the warning whether this can be ignored or not, in the markdown cell above, or disable this specific warning. 

2. GNIRS_IM_GRB120116A.ipynb
- In the Create Master Flat Field section, I see the following message in the 2nd cell: 
`“/bin/bash: showpars: command not found”.`
- Is there any particular reason for the markdown above the 2nd cell in boldface?

3. GASOI_NGC5128.ipynb
- In the Downing the data section, printing out the header gives me a long white space at the end
- In the Reduce Standard Star section, I got errors and warnings that look worrisome. Please check if they are typical. I copied some examples: 
```
“ERROR - Extensions have different shapes”
“WARNING - No REFCAT in S20170504S0114_sourcesDetected.fits - cannot calculate astrometry”
“WARNING - Downgrading fit to order 2 due to limited number of matches.”
```
- In the Reduce Science Images section, the execution of the cell takes too long (22 min for me). Is there a way to shorten the reduction time for example notebooks, e.g., by using a small portion of the image?

Let me know if any of these are unclear to you.

**Comment by bmerino95** [comment]
Hello again! I have finished adding the changes that @galaxyumi recommended. Several of the error/warning messages that were pointed out have been confirmed not to affect the final reduced product, and I found that several of those errors go away when using the latest version of DRAGONS (v3.2.0).

**Comment by rnikutta** [comment]
Hi @bmerino95 , here my comments and suggestions, following @jacquesalice 's and @galaxyumi 's reviews. Nice NBs, thanks for contributing them!

In this first round I have reviewed four NBs and the README file. I'll try to review the remaining three NBs this afternoon, but many comments from the first NBs (as I wrote down for Flamingos-2) apply to all NBs.)

**README.md**

- "The notebooks are set to open the DRAGONS kernel by default, but if this doesn’t happen, you can click ‘kernel’ at the top of the notebook, then ‘change notebook,’ and select DRAGONS (Py3.7) ..."

--> First, it's "then 'Change kernel...'". But second, it's even easier to change the kernel by clicking on the name of the current kernel in the top-right corner, and selecting a different kernel. You may want to give this instruction instead.

- The link "DRAGONS' Application Program Interface (API)" (https://dragons-recipe-system-users-manual.readthedocs.io/en/v2.1.1/appendices/full_api_example.html) gives me a 404 Not Found error.
- Alice already asked you to remove from the README file the section about installing DRAGONS and using conda... but it's still there(?)
- Please make sure to adjust the URLs in "Link to the Jupyter notebook" (for all NBs) to point to the `astro-datalab/notebooks-latest repo`, instead of `bmerino95/notebooks-latest`
- Please rename all capitalized "Flats, Darks, and Science frames" etc. to non-capitalized versions.
- For consistency, change "GMOS_longslit_WhiteDwarf.ipynb" (heading title) to "GMOS_Longslit_WhiteDwarf.ipynb"
- The link http://drforum.gemini.edu/ does not seem to be loading. Is the DR Forum up and running?
- The last section "Need help?" is specifically for people who run the original DRAGONS NBs on the local machines. It does not apply exactly for the NB versions hosted and run at Data Lab, right? Please check with Vini, and modify this section, so that it's clear in what cases users should contact Data Lab, and in what cases they should contact US-NGO.

**Flamingos2_Imaging_BrownDwarf NB**

- The title of a NB and the name of the ipynb file should have some resemblance. Please add "Brown Dwarf" somewhere to the title. Please fix for all NBs.
- In Introduction: the original Flamingos-2 tutorial is linked to three(!) times. Is this really necessary? Please combine the info in the paragraphs that you need to only link once. Fix in other NBs as well.
- The text ""A Study of the 450K Transition from T to Y Dwarf, and of the 350K Y Dwarfs" should be a link to https://archive.gemini.edu/programinfo/GS-2013B-Q-15 and hence you can then remove the text "More program information is given [here](https://archive.gemini.edu/programinfo/GS-2013B-Q-15)." Please fix in other NBs, too.
- In "Disclaimer and Attribution" the link to "Fitzpatrick et al.,  “The NOAO Data Laboratory: a conceptual overview” does not point to the paper, but a NB on Github instead.
- In "Importing Python libraries", remove `from __future__ import print_function` (this was needed in Python 2, but not in Python 3). Fix in other NBs as well.
- Group imports logically. For instance, both gempy imports should be next to each other, as should be both astropy imports. Please check for all NBs.
- In "About the dataset", change "This data comes from a Flamingos-2 observation program that observed a star" to "This data comes from a Flamingos-2 program that observed a star".
- The NB is about imaging of a Brown Dwarf. The final image displayed has a rather wide FOV. Would it be possible to add a marker pointing out the Brown Dwarf? And a few words of wisdom/conclusion? Without, the NB is a bit mechanical maybe..
- In the cleanup section, also add a single comment character to the empty row just below `%%bash`. This makes commenting in and out easier.

**GMOS_Imaging_Galaxy NB**

- This text "The data used for this tutorial is a dithered sequence on a starry field." is likely from the GMOS_Imaging_StarryField NB? Please fix for this NB.
- Some comments apply from the Flamongos-2 NB above.

**GMOS_Imaging_StarryField NB**

- Some comments from Flamingos-2 NB apply.
- In "" a link to the program description is missing in "The data used in this notebook example is GMOS I band imaging from the Gemini archive of a starry field from the Gemini North Hamamatsu CCD commissioning (Program: GN-2017A-SV-151)."
- The cleanup cell has an unnecessary white space between the comment character and the bash commands.

**GMOS_longslit_WhiteDwarf.ipynb NB**

- Some comments from Flamingos-2 NB apply.
- Rename the NB file from GMOS_longslit_WhiteDwarf.ipynb to GMOS_Longslit_WhiteDwarf.ipynb for consistency.
- "Because of the flexure, GMOS longslit flat field are not stacked." Should it be "flat fields"?
- As Yumi mentioned, this NB throws a lot of big scary ASDF deprecation warning ("WARNING: AstropyDeprecationWarning: ASDF functionality for astropy is being moved out of the astropy ...") Please deactivate it / filter it out it in the the Imports cell. (either astropy has a switch for that, or just use the 'warnings' module).
- The Science Observations reduction runs really a long time. Yumi asked about it too... No suggestions to shorten it? Fewer files maybe? At the very least, add a sentence before the cell warning the user that it will take "approximately N minutes".
- In "Display the 2D spectrum", would it be better to scale the image with something like LogNormal() or similar, to make more structure visible?
- In the cleanup cell:
```
# %%bash

# cp S20171022S0087_2D.fits final.fits
# cp S20171022S0087_1D.fits final_1D.fits
# rm -r calibrations
# rm *dat
# rm gmos_ls.list gmosls.log S2017*
# #rm bpm_20140601_gmos-s_Ham_22_full_12amp.fits
# #mv final.fits S20171022S0087_2D.fits
# #mv final_1D.fits S20171022S0087_1D.fits
```
add a comment char to the empty line, remove the whitespace betwee comment char and commands, either remove or un-comment the secondary comment char from the last three lines (depending what the intent is here..)

**Comment by rnikutta** [comment]
Round 2/2 of my comments @bmerino95 

BTW I ran all NBs on https://repgp02.datalab.noirlab.edu (you need to be on VPN), and had to change the kernel manually to "DRAGONS (Py3.7)" for each one. Recommend that at the end you execute each NB once on that machine, then regenerate the HTML files, and commit and push.

Please also update the `__version__` dates on each NB (at the end).

**GNIRS_Imaging_GammaRayBurst NB**

- Some comments from Flamingos-2 NB apply.
- Yumi also commented on it already: there are instances where the quotation characters are "slanted"; this happens for instance when one copy-pastes from a PDF etc. Example in this NB: `“Pass” setting in the “QA” column`. Replace with standard quotes: `"foobar"`.
- `showpars! N20120117S0014.fits associateSky` results in error `/bin/bash: line 1: showpars: command not found`. @galaxyumi already noted this, why wasn't it fixed?

**GSAOI_Imaging_EllipticalGalaxy NB**

- Some comments from Flamingos-2 NB apply.
- Yumi also mentioned this, but it hasn't been fixed: in some commented out cells like this one
```
# Check header of one raw science image
#tmp = fits.open("S20170505S0095.fits")
#tmp[0].header
```
The first line (a comment line) needs two consecutive comment characters lest it becomes 'code' (invalid code) if someone were to un-comment the entire cell). Please fix everywhere and in all NBs.
- In this same cell, when I un-comment it, the printed FITS header ends with a lot of extra whitespace. Again, Yumi already commented on it. Please fix. (and either have the cell, or don't. No need to have it but commented out.
- Under "Reduce standard star" the NB prints several times "ERROR - Extensions have different shapes". Is this ok? Also several times it prints "WARNING - No REFCAT in S20170504S0114_sourcesDetected.fits - cannot calculate astrometry".
- Under "Reduce science images", this sentence is not quite right grammatically I think: "You can find which file by its suffix (_flatCorrected or _skyCorrected)."

**NIRI_Imaging_Supernova NB**

- Some comments from Flamingos-2 NB apply.

**Comment by kareninysimba** [comment]
Hi Brian.

I took a peek at your notebooks, which are great! 
Only suggestion from a POV I took as a user who actually use them to reduce data, is the following:

* As mentioned in the final product section, these files will have the _stack.fits suffix, it should be mentioned for master dark, bpm, flat. At the end of each section ("Creating a master <calib>"), you should add something like 
*"The final master <calib> will have the name of the first <calib> file and the suffix _<calib>.fits"*
Why I consider this useful? When I was creating these files, I knew there will be new files, but not their names. Also, the display of files on the left side can be short and the image that was just created shows up at the end, so I would have to scroll to see it, or I could think nothing was produced. Also it can get blended in the list of files already there.
* The same for the corrected files (_flatCorrected.fits, _skyCorrected.fits)

In other words, whenever new images will be created, explain what their name are going to be.
It would be the same line repeated in each section. Only thing to change would be the string between '_' and '.fits'.

**Review by jacquesalice — APPROVED** [review]

**Comment by jacquesalice** [comment]
All looks good thank you @bmerino95 ! Merging now
